File: tools/mc-ref/gen_registry_nbt.py
#!/usr/bin/env python3
"""Convert vanilla datapack JSON → network NBT bytes for every synchronized registry entry."""
from __future__ import annotations

import json
import struct
import zipfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
JAR = Path(__file__).with_name("server-inner-26.2.jar")
OUT_BIN = ROOT / "crates" / "hyperion_server" / "src" / "network" / "join_data" / "registry_nbt.bin"
OUT_IDX = ROOT / "crates" / "hyperion_server" / "src" / "network" / "join_data" / "nbt_index.rs"

# Must match gen_join_data.py SYNC_FOLDERS + prioritization
SYNC = {
    "banner_pattern": "minecraft:banner_pattern",
    "cat_sound_variant": "minecraft:cat_sound_variant",
    "cat_variant": "minecraft:cat_variant",
    "chat_type": "minecraft:chat_type",
    "chicken_sound_variant": "minecraft:chicken_sound_variant",
    "chicken_variant": "minecraft:chicken_variant",
    "cow_sound_variant": "minecraft:cow_sound_variant",
    "cow_variant": "minecraft:cow_variant",
    "damage_type": "minecraft:damage_type",
    "dialog": "minecraft:dialog",
    "dimension_type": "minecraft:dimension_type",
    "enchantment": "minecraft:enchantment",
    "frog_variant": "minecraft:frog_variant",
    "instrument": "minecraft:instrument",
    "jukebox_song": "minecraft:jukebox_song",
    "painting_variant": "minecraft:painting_variant",
    "pig_sound_variant": "minecraft:pig_sound_variant",
    "pig_variant": "minecraft:pig_variant",
    "sulfur_cube_archetype": "minecraft:sulfur_cube_archetype",
    "test_environment": "minecraft:test_environment",
    "test_instance": "minecraft:test_instance",
    "timeline": "minecraft:timeline",
    "trim_material": "minecraft:trim_material",
    "trim_pattern": "minecraft:trim_pattern",
    "wolf_sound_variant": "minecraft:wolf_sound_variant",
    "wolf_variant": "minecraft:wolf_variant",
    "world_clock": "minecraft:world_clock",
    "zombie_nautilus_variant": "minecraft:zombie_nautilus_variant",
    "worldgen/biome": "minecraft:worldgen/biome",
}

TAG_END = 0
TAG_BYTE = 1
TAG_SHORT = 2
TAG_INT = 3
TAG_LONG = 4
TAG_FLOAT = 5
TAG_DOUBLE = 6
TAG_BYTE_ARRAY = 7
TAG_STRING = 8
TAG_LIST = 9
TAG_COMPOUND = 10
TAG_INT_ARRAY = 11
TAG_LONG_ARRAY = 12


def write_string(buf: bytearray, s: str) -> None:
    raw = s.encode("utf-8")
    if len(raw) > 65535:
        raise ValueError(f"string too long: {len(raw)}")
    buf.extend(struct.pack(">H", len(raw)))
    buf.extend(raw)


def tag_type_of(value) -> int:
    if isinstance(value, bool):
        return TAG_BYTE
    if isinstance(value, int):
        if -2147483648 <= value <= 2147483647:
            return TAG_INT
        return TAG_LONG
    if isinstance(value, float):
        return TAG_DOUBLE
    if isinstance(value, str):
        return TAG_STRING
    if isinstance(value, list):
        return TAG_LIST
    if isinstance(value, dict):
        return TAG_COMPOUND
    raise TypeError(f"unsupported JSON type: {type(value)}")


def write_payload(buf: bytearray, value) -> None:
    if isinstance(value, bool):
        buf.append(1 if value else 0)
    elif isinstance(value, int):
        if -2147483648 <= value <= 2147483647:
            buf.extend(struct.pack(">i", value))
        else:
            buf.extend(struct.pack(">q", value))
    elif isinstance(value, float):
        buf.extend(struct.pack(">d", value))
    elif isinstance(value, str):
        write_string(buf, value)
    elif isinstance(value, list):
        if not value:
            buf.append(TAG_END)
            buf.extend(struct.pack(">i", 0))
            return
        # Minecraft requires homogeneous lists; coerce by first element type
        et = tag_type_of(value[0])
        buf.append(et)
        buf.extend(struct.pack(">i", len(value)))
        for item in value:
            # light coercion for ints/floats mixed
            if et == TAG_DOUBLE and isinstance(item, int) and not isinstance(item, bool):
                item = float(item)
            if et == TAG_INT and isinstance(item, float) and item.is_integer():
                item = int(item)
            write_payload(buf, item)
    elif isinstance(value, dict):
        for k, v in value.items():
            if v is None:
                continue
            buf.append(tag_type_of(v))
            write_string(buf, k)
            write_payload(buf, v)
        buf.append(TAG_END)
    else:
        raise TypeError(type(value))


def encode_root_compound(obj: dict) -> bytes:
    """Network NBT: root compound with no name."""
    buf = bytearray()
    buf.append(TAG_COMPOUND)
    for k, v in obj.items():
        if v is None:
            continue
        buf.append(tag_type_of(v))
        write_string(buf, k)
        write_payload(buf, v)
    buf.append(TAG_END)
    return bytes(buf)


def list_entries(z: zipfile.ZipFile, folder: str) -> list[str]:
    prefix = f"data/minecraft/{folder}/"
    out = []
    for name in z.namelist():
        if name.startswith(prefix) and name.endswith(".json"):
            rest = name[len(prefix):-5]
            if "/" not in rest:
                out.append(rest)
    out.sort()
    return out


def prioritize(entries: list[str], first: str) -> list[str]:
    if first in entries:
        entries = [e for e in entries if e != first]
        entries.insert(0, first)
    return entries


def main() -> None:
    # Build blob: sequence of (registry_id, entry_path, nbt_bytes)
    # File format:
    #   u32be count
    #   repeated:
    #     u16be reg_len + reg_utf8
    #     u16be path_len + path_utf8
    #     u32be nbt_len + nbt
    records: list[tuple[str, str, bytes]] = []
    with zipfile.ZipFile(JAR) as z:
        for folder, reg_id in SYNC.items():
            entries = list_entries(z, folder)
            if reg_id == "minecraft:worldgen/biome":
                entries = prioritize(entries, "plains")
            if reg_id == "minecraft:dimension_type":
                entries = prioritize(entries, "overworld")
            for path in entries:
                jname = f"data/minecraft/{folder}/{path}.json"
                obj = json.loads(z.read(jname))
                if not isinstance(obj, dict):
                    raise SystemExit(f"expected object in {jname}")
                nbt = encode_root_compound(obj)
                records.append((reg_id, path, nbt))

    blob = bytearray()
    blob.extend(struct.pack(">I", len(records)))
    for reg_id, path, nbt in records:
        rb = reg_id.encode()
        pb = path.encode()
        blob.extend(struct.pack(">H", len(rb)))
        blob.extend(rb)
        blob.extend(struct.pack(">H", len(pb)))
        blob.extend(pb)
        blob.extend(struct.pack(">I", len(nbt)))
        blob.extend(nbt)

    OUT_BIN.parent.mkdir(parents=True, exist_ok=True)
    OUT_BIN.write_bytes(blob)
    print(f"wrote {OUT_BIN} records={len(records)} bytes={len(blob)}")

    # smoke: overworld nbt starts with 0x0a
    for reg_id, path, nbt in records:
        if reg_id.endswith("dimension_type") and path == "overworld":
            assert nbt[0] == 0x0A
            print("overworld nbt len", len(nbt))
            break


if __name__ == "__main__":
    main()

File: crates/hyperion_tools/src/bin/gen_registry_nbt.rs
//! Convert vanilla datapack JSON → packed network NBT for every sync registry entry.
//!
//! ```text
//! cargo run -p hyperion_tools --bin gen-registry-nbt
//! ```
//!
//! Output: `crates/hyperion_server/src/network/join_data/registry_nbt.bin`
//!
//! Blob format (big-endian):
//! ```text
//! u32 count
//! repeated:
//!   u16 reg_len + reg_utf8
//!   u16 path_len + path_utf8
//!   u32 nbt_len + nbt bytes (root compound, nameless)
//! ```

use std::io::Read;
use std::process::ExitCode;

use hyperion_protocol::{NbtTag, encode_compound_tag};
use hyperion_tools::{
    SYNC_FOLDERS, list_entries, open_server_jar, prioritize_registry, workspace_root, write_bytes,
};
use serde_json::Value;

fn main() -> ExitCode {
    match run() {
        Ok(()) => ExitCode::SUCCESS,
        Err(message) => {
            eprintln!("gen-registry-nbt: {message}");
            ExitCode::FAILURE
        }
    }
}

fn run() -> Result<(), String> {
    let root = workspace_root()?;
    let mut archive = open_server_jar(&root)?;
    let mut records: Vec<(String, String, Vec<u8>)> = Vec::new();

    for &(folder, reg_id) in SYNC_FOLDERS {
        let mut entries = list_entries(&mut archive, folder)?;
        prioritize_registry(reg_id, &mut entries);
        for path in entries {
            let jname = format!("data/minecraft/{folder}/{path}.json");
            let mut file = archive
                .by_name(&jname)
                .map_err(|e| format!("open {jname}: {e}"))?;
            let mut raw = String::new();
            file.read_to_string(&mut raw)
                .map_err(|e| format!("read {jname}: {e}"))?;
            let obj: Value =
                serde_json::from_str(&raw).map_err(|e| format!("parse {jname}: {e}"))?;
            let Value::Object(map) = obj else {
                return Err(format!("expected object in {jname}"));
            };
            let nbt = json_object_to_network_nbt(&map).map_err(|e| format!("nbt {jname}: {e}"))?;
            records.push((reg_id.to_owned(), path, nbt));
        }
    }

    let mut blob = Vec::new();
    blob.extend_from_slice(&(records.len() as u32).to_be_bytes());
    for (reg_id, path, nbt) in &records {
        let rb = reg_id.as_bytes();
        let pb = path.as_bytes();
        if rb.len() > u16::MAX as usize || pb.len() > u16::MAX as usize {
            return Err("registry or path string too long".to_owned());
        }
        blob.extend_from_slice(&(rb.len() as u16).to_be_bytes());
        blob.extend_from_slice(rb);
        blob.extend_from_slice(&(pb.len() as u16).to_be_bytes());
        blob.extend_from_slice(pb);
        blob.extend_from_slice(&(nbt.len() as u32).to_be_bytes());
        blob.extend_from_slice(nbt);
    }

    let out = root.join("crates/hyperion_server/src/network/join_data/registry_nbt.bin");
    write_bytes(&out, &blob)?;
    println!(
        "wrote {} records={} bytes={}",
        out.display(),
        records.len(),
        blob.len()
    );

    for (reg_id, path, nbt) in &records {
        if reg_id.ends_with("dimension_type") && path == "overworld" {
            if nbt.first() != Some(&0x0a) {
                return Err("overworld NBT must start with compound tag 0x0a".to_owned());
            }
            println!("overworld nbt len {}", nbt.len());
            break;
        }
    }
    Ok(())
}

fn json_object_to_network_nbt(map: &serde_json::Map<String, Value>) -> Result<Vec<u8>, String> {
    let entries = json_map_to_compound(map)?;
    encode_compound_tag(&entries).map_err(|e| e.to_string())
}

fn json_map_to_compound(
    map: &serde_json::Map<String, Value>,
) -> Result<Vec<(String, NbtTag)>, String> {
    let mut entries = Vec::with_capacity(map.len());
    for (k, v) in map {
        if v.is_null() {
            continue;
        }
        entries.push((k.clone(), json_to_nbt(v)?));
    }
    Ok(entries)
}

fn json_to_nbt(value: &Value) -> Result<NbtTag, String> {
    match value {
        Value::Null => Err("null is not a valid NBT value".to_owned()),
        Value::Bool(b) => Ok(NbtTag::Byte(if *b { 1 } else { 0 })),
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                if (i32::MIN as i64..=i32::MAX as i64).contains(&i) {
                    Ok(NbtTag::Int(i as i32))
                } else {
                    Ok(NbtTag::Long(i))
                }
            } else if let Some(f) = n.as_f64() {
                Ok(NbtTag::Double(f))
            } else {
                Err(format!("unsupported number {n}"))
            }
        }
        Value::String(s) => Ok(NbtTag::String(s.clone())),
        Value::Array(arr) => {
            if arr.is_empty() {
                return Ok(NbtTag::List(Vec::new()));
            }
            let et = tag_type_of(&arr[0])?;
            let mut items = Vec::with_capacity(arr.len());
            for item in arr {
                let coerced = coerce_for_list(item, et)?;
                items.push(coerced);
            }
            Ok(NbtTag::List(items))
        }
        Value::Object(map) => Ok(NbtTag::Compound(json_map_to_compound(map)?)),
    }
}

#[derive(Clone, Copy, PartialEq, Eq)]
enum JsonNbtKind {
    Byte,
    Int,
    Long,
    Double,
    String,
    List,
    Compound,
}

fn tag_type_of(value: &Value) -> Result<JsonNbtKind, String> {
    match value {
        Value::Bool(_) => Ok(JsonNbtKind::Byte),
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                if (i32::MIN as i64..=i32::MAX as i64).contains(&i) {
                    Ok(JsonNbtKind::Int)
                } else {
                    Ok(JsonNbtKind::Long)
                }
            } else {
                Ok(JsonNbtKind::Double)
            }
        }
        Value::String(_) => Ok(JsonNbtKind::String),
        Value::Array(_) => Ok(JsonNbtKind::List),
        Value::Object(_) => Ok(JsonNbtKind::Compound),
        Value::Null => Err("null list element".to_owned()),
    }
}

fn coerce_for_list(item: &Value, et: JsonNbtKind) -> Result<NbtTag, String> {
    match (et, item) {
        (JsonNbtKind::Double, Value::Number(n)) if n.as_i64().is_some() => {
            Ok(NbtTag::Double(n.as_i64().unwrap() as f64))
        }
        (JsonNbtKind::Int, Value::Number(n)) if n.as_f64().is_some_and(|f| f.fract() == 0.0) => {
            Ok(NbtTag::Int(n.as_f64().unwrap() as i32))
        }
        _ => json_to_nbt(item),
    }
}
